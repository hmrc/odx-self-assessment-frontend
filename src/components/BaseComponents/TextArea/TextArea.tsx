import React from 'react';
import PropTypes from 'prop-types';
import FormGroup, { makeErrorId, makeHintId } from '../FormGroup/FormGroup';
import HintTextComponent from '../../helpers/formatters/ParsedHtml';
import { useTranslation } from 'react-i18next';

export default function TextArea(props) {
  const {
    name,
    errorText,
    hintText,
    inputProps = {},
    expectedLength,
    id,
    onBlur,
    disabled,
    rows = 5,
    useCharacterCount = true // Add this prop
  } = props;
  const { t } = useTranslation();

  if (disabled) {
    return (
      <>
        {hintText && (
          <div id={makeHintId(name)} className='govuk-hint'>
            <HintTextComponent htmlString={hintText} />
          </div>
        )}
        <span className='govuk-body govuk-!-font-weight-bold read-only'>{inputProps.value}</span>
        <br />
      </>
    );
  }

  const describedByIDs = `${hintText ? makeHintId(name) : ''} ${
    errorText ? makeErrorId(name) : ''
  }`.trim();
  if (describedByIDs.length !== 0) {
    inputProps['aria-describedby'] = describedByIDs;
  }

  const remainingChars = expectedLength && expectedLength - inputProps.value?.length;
  const inputClasses = `govuk-textarea ${
    errorText || remainingChars < 0 ? 'govuk-textarea--error' : ''
  }`.trim();

  return (
    <FormGroup {...props} useCharacterCount={useCharacterCount}>
      <textarea
        className={inputClasses}
        {...inputProps}
        id={id}
        name={name}
        onBlur={onBlur}
        rows={rows}
        aria-describedby='with-hint-info'
      >
        {inputProps.value}
      </textarea>
      {expectedLength && (
        <div
          id='with-hint-info'
          className={`govuk-character-count__message govuk-character-count__status ${
            remainingChars < 0 ? 'govuk-error-message' : 'govuk-hint'
          }`}
        >
          {`${t('YOU_HAVE')}  ${Math.abs(remainingChars)} ${
            Math.abs(remainingChars) === 1 ? t('CHARACTER') : t('CHARACTERS')
          } ${remainingChars >= 0 ? t('REMAINING') : t('TOO_MANY')}`}
        </div>
      )}
    </FormGroup>
  );
}

TextArea.propTypes = {
  ...FormGroup.propTypes,
  name: PropTypes.string,
  maxLength: PropTypes.number,
  rows: PropTypes.number,
  inputProps: PropTypes.object,
  id: PropTypes.string,
  disabled: PropTypes.bool,
  onBlur: PropTypes.func,
  useCharacterCount: PropTypes.bool, // Add this prop type
  expectedLength: PropTypes.number
};
