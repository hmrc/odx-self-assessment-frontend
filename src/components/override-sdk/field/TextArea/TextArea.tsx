import React, { useContext, useEffect, useState } from 'react';
import GDSTextArea from '../../../BaseComponents/TextArea/TextArea';
import useIsOnlyField from '../../../helpers/hooks/QuestionDisplayHooks';
import ReadOnlyDisplay from '../../../BaseComponents/ReadOnlyDisplay/ReadOnlyDisplay';
import { registerNonEditableField } from '../../../helpers/hooks/QuestionDisplayHooks';
import GDSCheckAnswers from '../../../BaseComponents/CheckAnswer/index';
import { ReadOnlyDefaultFormContext } from '../../../helpers/HMRCAppContext';
import { checkStatus } from '../../../helpers/utils';

export default function TextArea(props) {
  const {
    getPConnect,
    value = '',
    placeholder,
    validatemessage,
    onChange,
    helperText,
    inputProps,
    fieldMetadata,
    readOnly,
    disabled,
    name,
    testId,
    configAlternateDesignSystem
  } = props;

  const { hasBeenWrapped } = useContext(ReadOnlyDefaultFormContext);

  const isComplexQuestionPage = PCore.getStoreValue('isComplexQuestionPage', '', 'app');

  const localizedVal = PCore.getLocaleUtils().getLocaleValue;
  const [errorMessage, setErrorMessage] = useState(localizedVal(validatemessage));

  registerNonEditableField(!!disabled);

  useEffect(() => {
    setErrorMessage(localizedVal(validatemessage));
  }, [validatemessage]);
  const thePConn = getPConnect();

  const propName = thePConn.getStateProps().value;
  const formattedPropertyName = name || propName?.split('.')?.pop();

  let label = props.label;
  const { isOnlyField, overrideLabel } = useIsOnlyField(props.displayOrder);
  if (isOnlyField && !readOnly) label = overrideLabel.trim() ? overrideLabel : label;

  const expectedLength = fieldMetadata?.expectedLength;
  const inprogressStatus = checkStatus();

  if (
    hasBeenWrapped &&
    configAlternateDesignSystem?.ShowChangeLink &&
    inprogressStatus === 'Open-InProgress'
  ) {
    return (
      <GDSCheckAnswers
        label={props.label}
        value={value}
        name={name}
        stepId={configAlternateDesignSystem.stepId}
        hiddenText={configAlternateDesignSystem.hiddenText}
        getPConnect={getPConnect}
        required={false}
        disabled={false}
        validatemessage=''
        onChange={undefined}
        readOnly={false}
        testId=''
        helperText={helperText}
        hideLabel={false}
        placeholder={placeholder}
      />
    );
  }

  if (readOnly) {
    return <ReadOnlyDisplay label={label} value={value} name={name} />;
  }

  const extraProps = { testProps: { 'data-test-id': testId } };

  const extraInputProps = { onChange, value };

  if (configAlternateDesignSystem?.autocomplete) {
    extraInputProps['autoComplete'] = configAlternateDesignSystem.autocomplete;
  } else {
    extraInputProps['autoComplete'] = 'off';
  }

  return (
    <GDSTextArea
      inputProps={{
        ...inputProps,
        ...extraInputProps
      }}
      hintText={helperText}
      extraLabelClasses={isComplexQuestionPage ? 'govuk-label--m' : ''}
      errorText={errorMessage}
      label={label}
      labelIsHeading={isOnlyField}
      name={formattedPropertyName}
      expectedLength={expectedLength}
      id={formattedPropertyName}
      {...extraProps}
      disabled={disabled || false}
    />
  );
}
