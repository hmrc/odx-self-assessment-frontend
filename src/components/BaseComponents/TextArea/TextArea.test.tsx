import React from 'react';
import { render, screen, fireEvent, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import TextArea from './TextArea';
// Note: running cleanup afterEach is done automatically for you in @testing-library/react@9.0.0 or higher
// unmount and cleanup DOM after the test is finished.

afterEach(cleanup);

jest.mock('@pega/auth/lib/sdk-auth-manager', () => ({ getSdkConfig: jest.fn() }));

jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: key => key
  })
}));

/**
 * SNAPHOTS
 */

describe('TextArea Component Snapshot', () => {
  const defaultProps = {
    name: 'test-textarea',
    id: 'test-textarea-id',
    inputProps: { value: 'Initial value' },
    expectedLength: 100,
    hintText: 'This is a hint',
    errorText: 'This is an error',
    onBlur: jest.fn()
  };

  it('matches the snapshot', () => {
    const { asFragment } = render(<TextArea {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('matches the snapshot when disabled', () => {
    const { asFragment } = render(<TextArea {...defaultProps} disabled />);
    expect(asFragment()).toMatchSnapshot();
  });
});

/**
 * ASSERTIONS
 */

describe('TextArea Component', () => {
  test('renders with hint text', () => {
    render(<TextArea name='test' hintText='This is a hint' />);
    expect(screen.getByText('This is a hint')).toBeInTheDocument();
  });

  test('displays remaining characters', () => {
    render(<TextArea name='test' expectedLength={10} inputProps={{ value: '12345' }} />);
    expect(screen.getByText('YOU_HAVE 5 CHARACTERS REMAINING')).toBeInTheDocument();
  });

  test('displays error message when characters exceed expectedLength', () => {
    render(<TextArea name='test' expectedLength={5} inputProps={{ value: '123456' }} />);
    expect(screen.getByText('YOU_HAVE 1 CHARACTER TOO_MANY')).toBeInTheDocument();
  });

  test('calls onBlur when textarea loses focus', () => {
    const handleBlur = jest.fn();
    render(<TextArea name='test' onBlur={handleBlur} />);
    fireEvent.blur(screen.getByRole('textbox'));
    expect(handleBlur).toHaveBeenCalledTimes(1);
  });

  test('renders character count when useCharacterCount is true', () => {
    render(
      <TextArea name='test' expectedLength={10} inputProps={{ value: '12345' }} useCharacterCount />
    );
    expect(screen.getByText('YOU_HAVE 5 CHARACTERS REMAINING')).toBeInTheDocument();
  });

  test('does not render character count when useCharacterCount is false', () => {
    render(
      <TextArea
        name='test'
        expectedLength={10}
        inputProps={{ defaultValue: '12345' }}
        useCharacterCount={false}
      />
    );
    expect(screen.queryByText('YOU_HAVE 5 CHARACTERS_REMAINING')).not.toBeInTheDocument();
  });

  test('adds govuk-character-count class when useCharacterCount is true', () => {
    const { container } = render(<TextArea name='test' useCharacterCount />);
    expect(container.querySelector('.govuk-character-count')).toBeInTheDocument();
  });

  test('does not add govuk-character-count class when useCharacterCount is false', () => {
    const { container } = render(<TextArea name='test' useCharacterCount={false} />);
    expect(container.querySelector('.govuk-character-count')).not.toBeInTheDocument();
  });

  test('adds error class when characters exceed expectedLength and textarea loses focus', () => {
    const { container } = render(
      <TextArea name='test' expectedLength={5} inputProps={{ value: '123456' }} />
    );

    // Simulate blur event
    fireEvent.blur(screen.getByRole('textbox'));

    // Check if the textarea has the error class
    const textarea = container.querySelector('textarea');
    expect(textarea).toHaveClass('govuk-textarea--error');
  });
});
