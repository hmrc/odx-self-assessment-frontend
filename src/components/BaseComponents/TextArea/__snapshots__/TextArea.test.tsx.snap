// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea Component Snapshot matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="govuk-form-group govuk-form-group--error  govuk-character-count"
  >
    <label
      class="govuk-label"
      for="test-textarea-id"
    />
    <div
      class="govuk-hint"
      id="test-textarea-hint"
    >
      This is a hint
    </div>
    <p
      class="govuk-error-message"
      id="test-textarea-error"
    >
      <span
        class="govuk-visually-hidden"
      >
        ERROR:
      </span>
      This is an error
    </p>
    <div>
      <textarea
        aria-describedby="with-hint-info"
        class="govuk-textarea govuk-textarea--error"
        id="test-textarea-id"
        name="test-textarea"
        rows="5"
      >
        Initial value
      </textarea>
      <div
        class="govuk-character-count__message govuk-character-count__status govuk-hint"
        id="with-hint-info"
      >
        YOU_HAVE  87 CHARACTERS REMAINING
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TextArea Component Snapshot matches the snapshot when disabled 1`] = `
<DocumentFragment>
  <div
    class="govuk-hint"
    id="test-textarea-hint"
  >
    This is a hint
  </div>
  <span
    class="govuk-body govuk-!-font-weight-bold read-only"
  >
    Initial value
  </span>
  <br />
</DocumentFragment>
`;
